R/cal-validate.R

```diff
@@ -0,0 +1,239 @@
+                                   parameters = NULL,
+                                   metrics = NULL,
+                                   summarize = TRUE,
+                                   ...) {
+  cal_validate(
+    rset = .data,
+    truth = {{ truth }},
+    estimate = {{ estimate }},
+    cal_function = "beta",
+    metrics = metrics,
+    summarize = summarize,
+    shape_params = shape_params,
+    location_params = location_params,
+    ...
+  )
+}
+
+cal_validate <- function(rset,
+                         truth = NULL,
+                         estimate = NULL,
+                         cal_function = NULL,
+                         metrics = NULL,
+                         summarize = TRUE,
+                         ...) {
+  if (is.null(cal_function)) rlang::abort("No calibration function provided")
+
+  if (is.null(metrics)) {
+    metrics <- yardstick::metric_set(
+      yardstick::brier_class
+    )
+  }
+
+  data_tr <- purrr::map(rset$splits, rsample::training)
+  data_as <- purrr::map(rset$splits, rsample::assessment)
+
+  if (cal_function == "logistic") {
+    cals <- purrr::map(
+      data_tr,
+      cal_estimate_logistic,
+      truth = {{ truth }},
+      estimate = {{ estimate }},
+      ...
+    )
+  }
+
+  if (cal_function == "isotonic") {
+    cals <- purrr::map(
+      data_tr,
+      cal_estimate_isotonic,
+      truth = {{ truth }},
+      estimate = {{ estimate }},
+      ...
+    )
+  }
+
+  if (cal_function == "beta") {
+    cals <- purrr::map(
+      data_tr,
+      cal_estimate_beta,
+      truth = {{ truth }},
+      estimate = {{ estimate }},
+      ...
+    )
+  }
+
+  estimate_col <- cals[[1]]$levels[[1]]
+
+  applied <- seq_along(data_as) %>%
+    map(
+      ~ {
+        val <- cal_apply(data_as[[.x]], cals[[.x]])
+        stats_after <- metrics(val, truth = {{ truth }}, estimate_col)
+        stats_before <- metrics(data_as[[.x]], truth = {{ truth }}, estimate_col)
+
+        list(
+          val = val,
+          stats_after = stats_after,
+          stats_before = stats_before
+        )
+      }
+    ) %>%
+    purrr::transpose()
+
+  ret <- dplyr::mutate(
+    rset,
+    calibration = cals,
+    validation = applied$val,
+    stats_after = applied$stats_after,
+    stats_before = applied$stats_before
+  )
+
+  if (summarize) ret <- summarize_validation(ret)
+
+  ret
+}
+
+#' @importFrom pillar type_sum
+#' @export
+type_sum.cal_binary <- function(x, ...) {
+  paste0(x$method, " [", x$rows, "]")
+}
+
+
+summarize_validation <- function(x) {
+  fs <- x$stats_after[[1]]
+
+  fs$.estimate <- NULL
+
+  seq_len(nrow(fs)) %>%
+    map(~ {
+      y <- .x
+      sa <- purrr::map_dbl(x$stats_after, ~ .x[y, ]$.estimate)
+      mean_sa <- mean(sa)
+      sb <- purrr::map_dbl(x$stats_before, ~ .x[y, ]$.estimate)
+      mean_sb <- mean(sb)
+      ret <- fs[y, ]
+      ret$.before <- mean_sb
+      ret$.after <- mean_sa
+      ret
+    }) %>%
+    dplyr::bind_rows()
```

Comment:
Maybe also add the `direction` column in too. Use `as_tibble(metrics) %>% select(metric, direction)` and merge that into the results?

Reply:
I wish I know why this happens 

```
> segment_logistic %>%
+     rsample::vfold_cv() %>%
+     cal_validate_isotonic(Class, metrics = metric_set(brier_class, roc_auc), summarize = FALSE)
#  10-fold cross-validation 
# A tibble: 10 × 6
   splits            id     calibration      validation         stats_after      stats_before
   <list>            <chr>  <list>           <list>             <list>           <list>      
 1 <split [909/101]> Fold01 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 2 <split [909/101]> Fold02 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 3 <split [909/101]> Fold03 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 4 <split [909/101]> Fold04 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 5 <split [909/101]> Fold05 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 6 <split [909/101]> Fold06 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 7 <split [909/101]> Fold07 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 8 <split [909/101]> Fold08 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
 9 <split [909/101]> Fold09 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
10 <split [909/101]> Fold10 <Isotonic [909]> <tibble [101 × 3]> <tibble [2 × 3]> <tibble>    
```

(`<tibble [2 × 3]>` vs  `<tibble>`) ¯\\_(⊙︿⊙)_/¯

Not that it is your problem.

Reply:
I think it is related to page width but still very frustrating for a developer.